File: tools/cogstore/jcs.py
"""RFC 8785 (JSON Canonicalization Scheme) — the subset cog-store catalogs use.

Catalogs are restricted to ASCII strings and integer numbers (protocol §7.1), where
RFC 8785 coincides exactly with
    json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False)

`canonical()` MUST produce bytes identical to
docs/protocol/testvectors/catalog.canonical.json — the package self-test asserts it,
so a regression here is caught immediately.

For general content (floats, non-ASCII) a full RFC 8785 implementation is required; the
guard below refuses anything outside the supported subset rather than emit wrong bytes.
"""
import json


def canonical(obj) -> bytes:
    """Return the RFC 8785 canonical UTF-8 bytes of `obj` (ASCII+integer subset)."""
    _assert_subset(obj)
    return json.dumps(obj, sort_keys=True, separators=(",", ":"),
                      ensure_ascii=False).encode("utf-8")


def _assert_subset(o, path="$"):
    # bool must be checked before int (bool is a subclass of int in Python).
    if isinstance(o, bool) or o is None or isinstance(o, int):
        return
    if isinstance(o, float):
        raise ValueError(f"{path}: float not allowed — use integers (protocol §7.1)")
    if isinstance(o, str):
        try:
            o.encode("ascii")
        except UnicodeEncodeError:
            raise ValueError(f"{path}: non-ASCII string outside this JCS subset")
        return
    if isinstance(o, dict):
        for k, v in o.items():
            if not isinstance(k, str):
                raise ValueError(f"{path}: non-string object key {k!r}")
            _assert_subset(v, f"{path}.{k}")
        return
    if isinstance(o, list):
        for i, v in enumerate(o):
            _assert_subset(v, f"{path}[{i}]")
        return
    raise ValueError(f"{path}: unsupported type {type(o).__name__}")

File: crates/gearbox/src/jcs.rs
//! RFC 8785 (JSON Canonicalization Scheme) for the cog-store subset.
//!
//! Catalogs are restricted to ASCII strings and integer numbers (protocol §7.1), so this
//! faithful subset implementation suffices and avoids a JCS-crate dependency. `canonical`
//! MUST reproduce `docs/protocol/testvectors/catalog.canonical.json` byte-for-byte — the
//! integration test asserts it. Non-integer numbers are rejected rather than emitted with
//! a guessed format.

use serde_json::Value;

/// Canonical UTF-8 bytes of `value` per RFC 8785 (ASCII + integer subset).
pub fn canonical(value: &Value) -> Result<Vec<u8>, String> {
    let mut out = Vec::new();
    write_value(value, &mut out)?;
    Ok(out)
}

fn write_value(v: &Value, out: &mut Vec<u8>) -> Result<(), String> {
    match v {
        Value::Null => out.extend_from_slice(b"null"),
        Value::Bool(true) => out.extend_from_slice(b"true"),
        Value::Bool(false) => out.extend_from_slice(b"false"),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                out.extend_from_slice(i.to_string().as_bytes());
            } else if let Some(u) = n.as_u64() {
                out.extend_from_slice(u.to_string().as_bytes());
            } else {
                return Err("non-integer number is outside the JCS subset (protocol §7.1)".into());
            }
        }
        Value::String(s) => write_string(s, out),
        Value::Array(items) => {
            out.push(b'[');
            for (i, item) in items.iter().enumerate() {
                if i > 0 {
                    out.push(b',');
                }
                write_value(item, out)?;
            }
            out.push(b']');
        }
        Value::Object(map) => {
            // RFC 8785: object members sorted by key as UTF-16 code units. Sort explicitly
            // so the result is independent of serde_json's map ordering feature.
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort_unstable_by(|a, b| a.encode_utf16().cmp(b.encode_utf16()));
            out.push(b'{');
            for (i, k) in keys.iter().enumerate() {
                if i > 0 {
                    out.push(b',');
                }
                write_string(k, out);
                out.push(b':');
                write_value(&map[k.as_str()], out)?;
            }
            out.push(b'}');
        }
    }
    Ok(())
}

fn write_string(s: &str, out: &mut Vec<u8>) {
    out.push(b'"');
    for c in s.chars() {
        match c {
            '"' => out.extend_from_slice(b"\\\""),
            '\\' => out.extend_from_slice(b"\\\\"),
            '\u{08}' => out.extend_from_slice(b"\\b"),
            '\u{09}' => out.extend_from_slice(b"\\t"),
            '\u{0A}' => out.extend_from_slice(b"\\n"),
            '\u{0C}' => out.extend_from_slice(b"\\f"),
            '\u{0D}' => out.extend_from_slice(b"\\r"),
            c if (c as u32) < 0x20 => {
                out.extend_from_slice(format!("\\u{:04x}", c as u32).as_bytes());
            }
            c => {
                let mut buf = [0u8; 4];
                out.extend_from_slice(c.encode_utf8(&mut buf).as_bytes());
            }
        }
    }
    out.push(b'"');
}
